### ISSUE TITLE
Possible parallel executing

### ISSUE THREAD

**Issue by ApmeM** [sergeyzwezdin/Hangfire.Mongo#42]
There is an issue in MongoDistributedLock file
```
do
{
    isLockedBySomeoneElse = database.DistributedLock
            .Find(Builders<DistributedLockDto>.Filter.Eq(_ => _.Resource, resource) &
                    Builders<DistributedLockDto>.Filter.Ne(_ => _.ClientId, _options.ClientId))
            .FirstOrDefault() != null;

    if (isFirstAttempt)
        isFirstAttempt = false;
    else
        Thread.Sleep((int)timeout.TotalMilliseconds / 10);                       <-=-=-=-=-=-=-
}
while (isLockedBySomeoneElse && (lockTimeoutTime >= DateTime.Now));
```
Are you sure it is good idea to sleep AFTER you realize that lock is free and going to exit a loop?
In this case if 2 jobs start within 1/20 of timeout interval - both will be executed in a parallel.
I think this sleep should be before 'Find'

Also in case of multiple enqueuing at once like this 
```
BackgroundJob.Enqueue(() => plainJob.ExecuteJob());
BackgroundJob.Enqueue(() => plainJob.ExecuteJob());
BackgroundJob.Enqueue(() => plainJob.ExecuteJob());
BackgroundJob.Enqueue(() => plainJob.ExecuteJob());
BackgroundJob.Enqueue(() => plainJob.ExecuteJob());
```

All of them can first check if lock is free in this loop, and then update each-others lock to own lock (so all threads will start executing, but only one record will be in a database).

To solve this when you performing update you need to check if it was actual update or insert and if it was an update - just revert this update and continue waiting. 

Usually this issue is solved by database unique index, but I know that mongo clusters cant create indexes correctly on all machines.

**Comment by MarLoe** [comment]
I have spotted a number of issues with the current implementation of MongoDistrubutedLock. This is one of them :)
Also the fact that testing the lock and actually acquiring the lock is not an atomic  operation is a huge issue.
So I have created a pull request:
https://github.com/sergeyzwezdin/Hangfire.Mongo/pull/40
If you want to try it out, look at:
https://github.com/MarLoe/Hangfire.Mongo
https://www.nuget.org/packages/Hangfire.Mongo.MarLoe/1.0.0

/Martin

**Comment by MarLoe** [comment]
I believe this issue is fixed by the merge of pull request #40
